ec2/utils.py

```diff
@@ -76,120 +215,267 @@ def get_instance_state(instance, ctx):
 
     try:
         reservations = ec2_client.get_all_reservations(instance_id)
-    except (EC2ResponseError, BotoServerError) as e:
-        raise NonRecoverableError('(Node: {0}): Error. '
+    except boto.exception.EC2ResponseError as e:
+        report_all_instances(ctx=ctx)
+        raise NonRecoverableError('Error. '
+                                  'Failed to get instance by id: '
+                                  'API returned: {0}.'
+                                  .format(str(e)))
+    except boto.exception.BotoServerError as e:
+        raise NonRecoverableError('Error. '
                                   'Failed to get instance by id: '
-                                  'API returned: {1}.'
-                                  .format(ctx.instance.id, e))
+                                  'API returned: {0}.'
+                                  .format(str(e)))
 
     instance = reservations[0].instances[0]
 
     return instance
 
 
-def get_instance_attribute(instance, attribute, timeout_length):
-    """ given the related instance object
-        the given variable can be retrieved and returned
-    """
-    timeout = time.time() + timeout_length
+def get_attached_keypair_id(ctx):
+    relationship_type = 'instance_connected_to_keypair'
 
-    while instance.update() != 'running':
-        time.sleep(5)
-        if time.time() > timeout:
-            raise NonRecoverableError('Timed out while attemting to get the '
-                                      'instance {0}. Timeout length: {1}.'
-                                      .format(attribute, timeout_length))
-    attribute = getattr(instance, attribute)
-    return attribute
+    kplist = get_target_aws_resource_id(relationship_type, ctx=ctx)
+    if kplist:
+        return kplist[0]
+    else:
+        return kplist
 
 
-def get_private_dns_name(instance, timeout_length):
-    """ returns the private_dns_name variable for a given instance
-    """
-    return get_instance_attribute(instance,
-                                  'private_dns_name', timeout_length)
+def get_attached_security_group_ids(ctx):
+    relationship_type = 'instance_connected_to_security_group'
 
+    return get_target_aws_resource_id(relationship_type, ctx=ctx)
 
-def get_public_dns_name(instance, timeout_length):
-    """ returns the public_dns_name variable for a given instance
+
+def get_target_aws_resource_id(relationship_type, ctx):
+    """ This loops through the relationships of type and returns
+        targets of those relationships.
     """
-    return get_instance_attribute(instance,
-                                  'public_dns_name', timeout_length)
 
+    ids = []
 
-def get_private_ip_address(instance, timeout_length):
-    """ returns the private_ip_address variable for a given instance
-    """
-    return get_instance_attribute(instance,
-                                  'private_ip_address', timeout_length)
+    if not hasattr(ctx.instance, 'relationships'):
```

Comment:
changed to use getattr(ctx.instance, 'relationships', []) so that if an empty list is returned from the value of relationships then it's skipped;